src/bpftrace.cpp

```diff
@@ -1264,14 +1270,30 @@ int BPFtrace::run(BpfBytecode bytecode)
   if (run_special_probe("END_trigger", bytecode_, END_trigger))
     return -1;
 
-  poll_perf_events(/* drain */ true);
+  poll_output(/* drain */ true);
 
-  // Calls perf_reader_free() on all open perf buffers.
-  open_perf_buffers_.clear();
+  teardown_output();
 
   return 0;
 }
 
+int BPFtrace::setup_output()
+{
+  // when ringbuf is available, setup ringbuf for build-ins like printf, cat.
+  if (feature_->has_map_ringbuf())
+  {
+    int err = setup_ringbuf();
+    if (err)
+      return err;
+  }
+  // when ringbuf is unavailable or built-in skboutput is used, setup perf_event
+  if (!feature_->has_map_ringbuf() || resources.needs_perf_event_map)
```

Comment:
How about wrapping `!feature_->has_map_ringbuf() || resources.needs_perf_event_map` and `feature_->has_map_ringbuf()` into two helper functions instead? I see these conditionals used in a couple places and it would make it less error prone and easier to change to have a wrapper.